src/Lykke.Numerics.Money/Linq/Money.Average.cs

```diff
@@ -0,0 +1,102 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Lykke.Numerics;
+
+using static Lykke.Linq.ThrowHelper;
+
+namespace Lykke.Linq
+{
+    public static partial class MoneyEnumerable
+    {
+        public static Money Average(
+            this IEnumerable<Money> source)
+        {
+            if (source == null)
+            {
+                throw SourceArgumentNullException();
+            }
+
+            using (var enumerator = source.GetEnumerator())
+            {
+                if (!enumerator.MoveNext())
+                {
+                    throw SequenceContainsNoElementsException();
+                }
+
+                var sum = enumerator.Current;
+                var count = 1UL;
+                
+                checked
+                {
+                    while (enumerator.MoveNext())
+                    {
+                        sum += enumerator.Current;
+                        count++;
+                    }
+                }
+
+                return sum / count;
+            }
+        }
+
+        public static Money? Average(
+            this IEnumerable<Money?> source)
+        {
+            if (source == null)
+            {
+                throw SourceArgumentNullException();
+            }
+
+            using (var enumerator = source.GetEnumerator())
+            {
+                while (enumerator.MoveNext())
+                {
+                    var current = enumerator.Current;
+                    
+                    if (current.HasValue)
+                    {
+                        var sum = current.GetValueOrDefault();
```

Comment:
Here and many times below, why not just Value? We've already checked HasValue above

Reply:
Fixed for Average, but it's necessary other methods